docs/design/object-locking.md

```diff
@@ -0,0 +1,207 @@
+evaluated against the `retain-until-date` set on the object and the client will
+receive an `Access Denied` error if the current date and time is less than the
+`retain-until-date` set on the object.
+When the current date and time exceeds the `retain-until-date`, deletes on the
+object version are automatically allowed. There is no cleanup action to remove
+the `retain-until-date` set on the object version's metadata once the
+`retain-until-date` expires.
+DELETE object requests without a version id result in creation of delete markers
+on top of the object version, even if the object has a lock configuration set
+and the `retain-until-date` is current.
+
+#### Lifecycle actions
+
+Lifecycle jobs can create delete markers on the object but cannot permanently
+delete an object version that is locked until the lock expires.
+
+#### Replication
+
+The lock configuration on the object version in a source bucket is copied over
+to the destination bucket only if object lock is enabled on the destination
+bucket. Otherwise, the lock is ignored in the destination bucket.
+
+## APIs covering S3 Object lock
+
+* **Put Bucket** - extend the bucket creation API to include configuration for
+  enabling object lock on the bucket
+
+  **Note:** Versioning is automatically enabled on buckets that have object lock
+  enabled as part of a PUT Bucket request
+
+* **Put Object** - extend put object API to parse `x-amz-object-lock-mode`,
+  `x-amz-object-lock-retain-until-date`, `x-amz-object-lock-legal-hold`
+  headers and store the configuration on an object.
+
+* **Copy Object** - extend the API to accept the same lock configuration
+  headers as the PUT Object request
+
+* **Create Multipart Upload** - extend the API to accept the same lock
+  configuration headers as the PUT Object request
+
+* **Put Object Lock Configuration** - allows setting a default lock
+  configuration for objects that are going to be stored in the bucket. This
+  request is accepted only on buckets that have object lock enabled.
+
+* **Get Object Lock Configuration** - gets the object lock configuration set on
+  the bucket metadata
+
+* **Put Object Retention** - sets the retention mode/period configuration on a
+  object version
+
+* **Get Object Retention** - gets the retention mode/period configuration set
+  on a object version
+
+* **Put Object Legal Hold** - sets legal hold configuration on an object version
+
+* **Get Object Legal Hold** - gets the legal hold status for an object version
+
+## Deviations
+
+The following hightlights any deviation from AWS implementation of the feature
+
+- `token`
+
+  AWS allows setting object lock on existing buckets by asking the user to
+  contact support, the support then gives a `token` that can be used to set lock
+  on the specific bucket. Cloudserver's implementation differs in the way that
+  we will use a tool to enable object lock by modifying the bucket's configuration
```

Comment:
What kind of a tool does this referring to, is there any code change that needs to be done to allow setting object lock on an existing bucket? (based on our current implementation we're returning `Object Lock configuration cannot be enabled on existing buckets`)

Reply:
@rahulreddy for visibility

Reply:
@ilkescality  Yes, I will add more info on the tool. It can be a node.js script in s3utils that allows TS to bypass the S3API and set the object lock flag on buckets.

Reply:
Sounds good.